### PR TITLE
M1 compatibility

### PR DESCRIPTION
**Pull request by kevindew**
This PR tries to pull together all the fixes we know are needed for using GOV.UK docker on an M1 mac using information from: https://github.com/alphagov/govuk-docker/issues/561

~~This is marked as a draft as I'm not sure if we'll merge these things in directly, I expect we'd probably do some more testing and then consider adding some explanatory comments to code.~~

I've done some testing now and other M1 mac users are making use of this branch. I've also confirmed it allows database restores for MySQL, MongoDB and Elasticsearch.

### CODE REVIEW

File: docker-compose.yml
Comment (Tetrino) [inline docker-compose.yml:45]: For anyone who uses this branch, you'll need to make sure you use Docker's debug feature to purge all the images before you attempt using MySQL. Without doing so, you will have odd behaviour in regards to Whitehall connectivity and replicating data.

**Comment by DilwoarH** [comment]
Just tested Whitehall with this, ran fine 👍

File: docker-compose.yml
Comment (kevindew) [inline docker-compose.yml:42]: This command seems to be causing problems with Mongoid on applications.

On Slack people have reported that you can't run `make publisher` and we seem to have traced it to this change.

E.g:

```
$ git checkout m1-compatibility~1; govuk-docker down; govuk-docker run publisher-lite bundle exec rails db:setup
HEAD is now at b5b94d7 Use elasticsearch-7 for for M1 compatibility
docker-compose -f [...] down
[+] Running 4/3
 ⠿ Container govuk-docker_publisher-lite_run_fbbd6effc69e  Removed                                                                                                                                                                                                                                                                                                     0.0s
 ⠿ Container govuk-docker-mongo-3.6-1                      Removed                                                                                                                                                                                                                                                                                                     0.3s
 ⠿ Container govuk-docker-redis-1                          Removed                                                                                                                                                                                                                                                                                                     0.2s
 ⠿ Network govuk-docker_default                            Removed                                                                                                                                                                                                                                                                                                     0.1s
docker-compose -f [...] run publisher-lite bundle exec rails db:setup
[+] Running 1/1
[+] Running 3/3k-docker_default  Created                                                                                                                                                                                                                                                                                                                               0.0s
 ⠿ Network govuk-docker_default        Created                                                                                                                                                                                                                                                                                                                         0.0s
 ⠿ Container govuk-docker-redis-1      Created                                                                                                                                                                                                                                                                                                                         0.1s
 ⠿ Container govuk-docker-mongo-3.6-1  Created                                                                                                                                                                                                                                                                                                                         0.1s
[+] Running 2/2
 ⠿ Container govuk-docker-redis-1      Started                                                                                                                                                                                                                                                                                                                         0.3s
 ⠿ Container govuk-docker-mongo-3.6-1  Started                                                                                                                                                                                                                                                                                                                         0.4s
** Invoke db:setup (first_time)
** Invoke db:create (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute db:create
** Invoke db:mongoid:create_indexes (first_time)
** Invoke environment
** Invoke db:mongoid:load_models (first_time)
** Execute db:mongoid:load_models
** Execute db:mongoid:create_indexes
** Invoke db:seed (first_time)
** Invoke environment
** Execute db:seed
** Execute db:setup
```

vs:

```
$ git checkout m1-compatibility; govuk-docker down; govuk-docker run publisher-lite bundle exec rails db:setup
Previous HEAD position was b5b94d7 Use elasticsearch-7 for for M1 compatibility
Switched to branch 'm1-compatibility'
Your branch is up to date with 'origin/m1-compatibility'.
docker-compose -f [...] down
[+] Running 4/3
 ⠿ Container govuk-docker_publisher-lite_run_c698634de2a2  Removed                                                                                                                                                                                                                                                                                                     0.0s
 ⠿ Container govuk-docker-mongo-3.6-1                      Removed                                                                                                                                                                                                                                                                                                     0.3s
 ⠿ Container govuk-docker-redis-1                          Removed                                                                                                                                                                                                                                                                                                     0.2s
 ⠿ Network govuk-docker_default                            Removed                                                                                                                                                                                                                                                                                                     0.1s
docker-compose -f [...] run publisher-lite bundle exec rails db:setup
[+] Running 1/1
[+] Running 3/3k-docker_default  Created                                                                                                                                                                                                                                                                                                                               0.0s
 ⠿ Network govuk-docker_default        Created                                                                                                                                                                                                                                                                                                                         0.0s
 ⠿ Container govuk-docker-redis-1      Created                                                                                                                                                                                                                                                                                                                         0.1s
 ⠿ Container govuk-docker-mongo-3.6-1  Created                                                                                                                                                                                                                                                                                                                         0.1s
[+] Running 2/2
 ⠿ Container govuk-docker-mongo-3.6-1  Started                                                                                                                                                                                                                                                                                                                         0.4s
 ⠿ Container govuk-docker-redis-1      Started                                                                                                                                                                                                                                                                                                                         0.3s
** Invoke db:setup (first_time)
** Invoke db:create (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute db:create
** Invoke db:mongoid:create_indexes (first_time)
** Invoke environment
** Invoke db:mongoid:load_models (first_time)
** Execute db:mongoid:load_models
** Execute db:mongoid:create_indexes
rails aborted!
Mongo::Error::NoServerAvailable: No primary server is available in cluster: #<Cluster topology=Unknown[mongo-3.6:27017] servers=[#<Server address=mongo-3.6:27017 GHOST>]> with timeout=30, LT=0.015

Tasks: TOP => db:setup => db:mongoid:create_indexes
(See full trace by running task with --trace)
```

Comment (Tetrino) [reply docker-compose.yml:42]: Oddly `make publisher` has been working fine for me on this branch, but that is after tearing down my entire Docker environment in debug tools.

Comment (DilwoarH) [reply docker-compose.yml:42]: I get this issue too now: 

```
Mongo::Error::NoServerAvailable: No primary server is available in cluster: #<Cluster topology=Unknown[mongo-3.6:27017] servers=[#<Server address=mongo-3.6:27017 GHOST>]> with timeout=30, LT=0.015
```

Comment (kevindew) [reply docker-compose.yml]: @KludgeKML since this is causing problems. I've pushed this branch with b7110cb removed from it.

I've pushed up a new branch `m1-compatibility-keith` that still has it on it.

Comment (DilwoarH) [reply docker-compose.yml]: Why do we need it to replicate locally?

**Comment by kevindew** [comment]
> So, in short (because this is something I've been on a deep dive into recently): Router currently uses a feature called OpTime to find out whether it has to refresh its internal representation of the routing tree. The way Router is written it requires a replicaset (the optime calls basically open up the replicaset, dig down into it to find the primary, then use that as the check). The replicaset can be only 1 service (the primary), but it has to be there - router doesn't have any fallback. So if you're using router in a way on your local machine that needs you never to update the routes, it would work (although it would be chucking out warning messages about not finding the replica primary). If you wanted to update routes, or you wanted to run the router test suite, it would fail without the replicaset. At the moment I'm working on moving router in production to a DocumentDB database, which will require me to get rid of the replicaset requirement. But it's not quite done yet (and has just had a week+ delay due to me getting covid).

Do hope you're feeling better now Keith

Hmm I see, so for that aspect of router to work it needs to be running as a replica set. Whereas for all other apps they don't care if replicaset or not?

From what I can tell just starting mongo with `--replSet <replica-set-name>` doesn't seem to be enough anymore. You need to also run `rs.initiate()` in MongoDB directly to start the replicaset (I imagine there's perhaps other ways to do this, such as config files) - so it seems quite the pain.

Do you think we need to wait until(/if) we resolve this before we merge this PR for M1 compatibility?

I feel like, aside from this, all the other bits work.

**Comment by KludgeKML** [comment]
Short story: I think you should merge this for the moment and we keep a separate branch for working on router until I can update router enough that that's unnecessary.

Long story: I would say that this affects:
- people actively working on router (AFAIK only me + occasionally some people doing dependabot version bumps)
- people who want to test things locally that require route creation going all the way through to router 

People in group 1 won't be able to run the test suite locally. But it'll still run on Jenkins, and it can be made to run locally with the small change to their local govuk-docker. So I think they will live with that.

People in group 2 won't see routes get updated locally. But - that might not be a thing people care too much about in local dev? I'm also not 100% sure whether this is true, either - is the endpoint on router that triggers an update completely unused now? If some things still use it, maybe that is fine. You have a better idea of how frequent this need comes up, but I guess maybe not a lot?

Either way, the impact of merging this as is seems minimal. It might only ever affect me, and I already know how to work around it. So I'm happy for this to get merged.

**Review by KludgeKML — APPROVED** [review]
Looks good to me.

File: docker-compose.yml
Comment (KludgeKML) [reply docker-compose.yml:48]: Marking this as resolved for the moment, I'll make a new branch specifically for router development until I can resolve this.

File: docker-compose.yml
Comment (KludgeKML) [reply docker-compose.yml]: It's necessary for router development (but not for router work). Resolution is we merge this, and the handful of people doing active router development live with a small workaround until we can get router updated (coming soon).

**Pull request merged**
